File: Data Structure/Lab 8/Duplicate.c
//Write a program to remove the duplicates nodes from given sorted Linked List. 
//Input: 1 → 1 → 6 → 13 → 13 → 13 → 27 → 27 
// Output: 1 → 6 → 13 → 27

#include<stdio.h>
#include<stdlib.h>

struct Node
{
    int info;
    struct Node *link;
};

struct Node *First;

struct Node *insertatFirst(int x){

    struct Node *newNode = (struct Node*)malloc(sizeof (struct Node));
    newNode -> info = x;
    newNode -> link = First; 

    First = newNode;

    return First;
}

void display(struct Node *first){

    if(first == NULL){
        printf("Empty List1");
        return;
    }

    struct Node *save = first;

    while(save != NULL){
        printf("%d ->",save->info);
        save = save -> link;
    }
    printf("NULL\n");
}

void removeDuplicate(){

    if(First == NULL){
        printf("List is empty!");
        return;
    }

    struct Node *save = First;

    while(save != NULL && save->link != NULL){
        if(save->info == save->link->info){
            struct Node *temp = save->link;
            save->link = save->link->link;
            free(temp);
        }
        else{
            save = save->link;
        }
    }

}

void main(){

    insertatFirst(27);
    insertatFirst(27);
    insertatFirst(13);
    insertatFirst(13);
    insertatFirst(13);
    insertatFirst(6);
    insertatFirst(1);
    insertatFirst(1);

    display(First);

    removeDuplicate();
    
    display(First);

}

File: Data Structure(Java)/Lab 6/Duplicate.java
// Write a program to remove the duplicates nodes from given sorted Linked List. 
// Input: 1 → 1 → 6 → 13 → 13 → 13 → 27 → 27 
// Output: 1 → 6 → 13 → 27 

class Node{
    int info;
    Node link;

    Node(int info){
        this.info = info;
        this.link = null;;
    }
}

public class Duplicate {

    static Node first = null;
    public static void main(String[] args) {

        insertAtFirst(27);
        insertAtFirst(27);
        insertAtFirst(13);
        insertAtFirst(13); 
        insertAtFirst(13);
        insertAtFirst(6);   
        insertAtFirst(1);
        insertAtFirst(1);

        displayNodes();

        removeDuplicate();

        displayNodes();
    }

    static void insertAtFirst(int x) {
        Node newnode = new Node(x);
        newnode.link = first;
        first = newnode;
    }

    static void displayNodes() {
        if (first == null) {
            System.out.println("List is empty.");
            return;
        }
        Node save = first;
        while (save != null) {
            System.out.print(save.info + " -> ");
            save = save.link;
        }
        System.out.println("null");
    }

    static void removeDuplicate(){
        if(first == null){
            System.out.println("Empty List !!");
            return;
        }

        Node save = first;

        while(save != null && save.link != null){
            if(save.info == save.link.info){
                // Node temp = save.link;
                save.link = save.link.link;
            }
            else{
                save = save.link;
            }
        }
    }
}
